tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/RunProcessor.cs

```diff
@@ -0,0 +1,243 @@
+
+                var runUriPath = runUri.AbsolutePath;
+                var runUriPathSegments = runUri.PathAndQuery.Split("/");
+                var organization = runUriPathSegments[1];
+                var projectGuid = Guid.Parse(runUriPathSegments[2]);
+                var pipelineId = int.Parse(runUriPathSegments[5]);
+                var runId = int.Parse(runUriPathSegments[7]);
+
+                var azureDevOpsPersonalAccessToken = await GetAzureDevOpsPersonalAccessTokenAsync();
+                var credentials = new VssBasicCredential("nobody", azureDevOpsPersonalAccessToken);
+
+                var baseUri = new Uri($"https://dev.azure.com/{organization}");
+                var connection = new VssConnection(baseUri, credentials);
+
+                var buildClient = connection.GetClient<BuildHttpClient>();
+                var projectClient = connection.GetClient<ProjectHttpClient>();
+                var organizationClient = connection.GetClient<OrganizationHttpClient>();
+
+                var build = await buildClient.GetBuildAsync(projectGuid, runId);
+                var project = await projectClient.GetProject(projectGuid.ToString());
+                var pipeline = await buildClient.GetDefinitionAsync(project.Id, build.Definition.Id);
+                var timeline = await buildClient.GetBuildTimelineAsync(projectGuid, runId);
+
+                double agentDurationInSeconds = 0;
+                double queueDurationInSeconds = 0;
+
+                if (timeline != null)
+                {
+                    agentDurationInSeconds = (from record in timeline.Records
+                                              where record.RecordType == "Task"
+                                              where (record.Name == "Initialize job" || record.Name == "Finalize Job") // Love consistency in capitalization!
+                                              group record by record.ParentId into job
+                                              let jobStartTime = job.Min(jobRecord => jobRecord.StartTime)
+                                              let jobFinishTime = job.Max(jobRecord => jobRecord.FinishTime)
+                                              let agentDuration = jobFinishTime - jobStartTime
+                                              select agentDuration.Value.TotalSeconds).Sum();
+
+                    queueDurationInSeconds = (from taskRecord in timeline.Records
+                                              where taskRecord.RecordType == "Task"
+                                              where taskRecord.Name == "Initialize job"
+                                              join jobRecord in timeline.Records on taskRecord.ParentId equals jobRecord.Id
+                                              let queueDuration = taskRecord.StartTime - jobRecord.StartTime
+                                              select queueDuration.Value.TotalSeconds).Sum();
+                }
+
+                var run = new Run()
```

Comment:
Is it too much data to simply capture all the timeline records?

Reply:
I plan to do this for the next level of analysis (possibly including logs). But I wanted to get this in first and see how it goes before I start throwing that much data at it.

Reply:
For logs I was considering having one record per line of the logs, but that would be a huge amount of data to collect.